### PR TITLE
Feature/reference extract

### PR DESCRIPTION
**Pull request by GrahamMWillis**
(1) Add applicant and ecologist organisations to the model and the Open API definition. 
(2) Added a new package: refdata-extract-processor. This is expected to run at a slower cadence than the application-extract-processor. It extracts the global option sets and other reference data and make them available via the API. (New tables have been added added to postgres see [packages/database-model/src/sequelize-model.js ](url)
(3) Add lookup functionality to the transformation - the consumer specifies a named field of the reference data rather than the id (the lookups use data that is streamed directly from Power Apps (and cached in REDIS for 1 day). This occurs at transformation time in the queue consumer and there is no database action (since the qc relies on Power Apps availability). 
(4) Various enhancements to the model logic which is described in [packages/powerapps-lib/src/application/model/sdds-applications.js](url)

### CODE REVIEW

File: packages/refdata-extract-processor/package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  "private": false,
+  "publishConfig": {
+    "access": "public"
+  },
+  "engines": {
+    "node": ">=16.13.0"
+  },
+  "scripts": {
+    "start": "node src/processor.js",
+    "test": "echo \"Error: run tests from root\" && exit 1"
```
Comment (timtamimi) [inline packages/refdata-extract-processor/package.json:16]: Bonus brownie points if you can modify the 'test' command or remove it, but that's just for tidiness. No biggie.

**Review by timtamimi — CHANGES_REQUESTED** [review]
A few minor comments.

File: packages/api/src/handlers/application/application-section/put-section.js [hunk 2]
```diff
@@ -18,19 +18,19 @@ export default async (context, req, h) => {
     const sequelize = SEQUELIZE.getSequelize()
 
     const [, updatedApplication] = await models.applications.update({
-      application: sequelize.fn('jsonb_set', sequelize.col('application'), '{applicant}', JSON.stringify(req.payload), true)
+      application: sequelize.fn('jsonb_set', sequelize.col('application'), `{${section}}`, JSON.stringify(req.payload), true)
```
[the remark this answers is not in the document]
Comment (GrahamMWillis) [reply packages/api/src/handlers/application/application-section/put-section.js:21]: It is actually JSON that is required, I could change to JSON.stringify({section})
Ignore for now - I may revisit

File: packages/database-model/src/sequelize-model.js [hunk 3]
```diff
@@ -51,8 +34,32 @@ const createModels = async () => {
     ]
   })
 
+  models.applicationTypes = await sequelize.define('application-types', {
+    id: { type: DataTypes.UUID, primaryKey: true },
+    json: { type: DataTypes.JSONB }
```
[the remark this answers is not in the document]
Comment (GrahamMWillis) [reply packages/database-model/src/sequelize-model.js:39]: The id here is actually the PowerApps primary key. Currently this is informational as the actual lookup is done  in the transformation. Will extend at some point.

File: packages/powerapps-lib/src/application/model/sdds-applications.js [hunk 4]
```diff
@@ -1,30 +1,169 @@
...
 
 export const sddsApplications = {
   targetEntity: 'sdds_applications',
   targetKey: 'sdds_applicationid',
   targetFields: {
+    sdds_applicationnumber: { // Place holder - needs  further discussion
+      srcFunc: () => crypto.randomBytes(3).toString('hex').toUpperCase()
+    },
+
     sdds_descriptionofproposal: {
```
[the remark this answers is not in the document]
Comment (GrahamMWillis) [reply packages/powerapps-lib/src/application/model/sdds-applications.js:85]: These are the names in power-apps, so its a reflection of that. I could I suppose relegate the Power Apps identifier to a  field but its quite a lot of work to change all of those regression functions so if you don't mind I will leave for now.

File: packages/powerapps-lib/src/model/transformer.js [hunk 5]
```diff
@@ -11,32 +11,70 @@ const { default: has } = _has
...
 }
 
+async function powerAppsObjectBuildByFieldFunction (fields, field, src, obj) {
+  if (fields[field].bind) {
+    const id = await fields[field].srcFunc(src)
+    if (id) {
+      const data = `/${fields[field].bind}(${id})`
+      Object.assign(obj, { [`${field}@odata.bind`]: data })
+    } else {
+      console.log(`WARNING: expected bound relation not found: ${fields[field].bind}`)
```
[the remark this answers is not in the document]
Comment (GrahamMWillis) [reply packages/powerapps-lib/src/model/transformer.js:43]: According to the node documentation:

'The console.warn() function is an alias for console.error()'

I only want to raise an error if its an actual error and not an _unexpected data possibility_

File: packages/powerapps-lib/src/model/transformer.js [hunk 6]
```diff
@@ -45,28 +83,68 @@ export const localObjectBuilder = (node, paObj, obj = {}, keys = {}) => {
...
+  res.forEach(r => set(obj, r.srcPath, r.value))
+}
+
+async function apiObjectBuilderTraverse (node, nodeName, paObj, obj, keys) {
   for (const next in node[nodeName]?.relationships) {
     const nn = node[nodeName].relationships[next]
     const key = nn.fk.replace('@odata.bind', '')
     if (paObj[key]) {
-      localObjectBuilder({ [next]: nn }, paObj[key], obj, keys)
+      await apiObjectBuilder({ [next]: nn }, paObj[key], obj, keys)
```
[the remark this answers is not in the document]
Comment (GrahamMWillis) [reply packages/powerapps-lib/src/model/transformer.js:129]: Yes its required, the object build has the ability to call an async function

File: packages/refdata-extract-processor/package.json (hunk 1 above)
Comment (GrahamMWillis) [reply packages/refdata-extract-processor/package.json:16]: Sounds like a last thing Friday job! I will bear it in mind.

**Pull request merged**